// @abgov/adsp-service-sdk/src/tenant/tenantService.js
"use strict";
var _TenantServiceImpl_tenants, _TenantServiceImpl_tenantNames, _TenantServiceImpl_tenantRealms, _TenantServiceImpl_tryRetrieveTenants;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TenantServiceImpl = void 0;
const tslib_1 = require("tslib");
const axios_1 = require("axios");
const NodeCache = require("node-cache");
// eslint-disable-next-line @typescript-eslint/no-var-requires
const retry = require('promise-retry');
const utils_1 = require("../utils");
class TenantServiceImpl {
    constructor(logger, directory, tokenProvider, preload = true) {
        this.logger = logger;
        this.directory = directory;
        this.tokenProvider = tokenProvider;
        this.LOG_CONTEXT = { context: 'TenantService' };
        _TenantServiceImpl_tenants.set(this, new NodeCache({
            stdTTL: 36000,
            useClones: false,
        }));
        _TenantServiceImpl_tenantNames.set(this, {});
        _TenantServiceImpl_tenantRealms.set(this, {});
        _TenantServiceImpl_tryRetrieveTenants.set(this, (requestUrl, count, criteria) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            var _a;
            this.logger.debug(`Try ${count}: retrieving tenants from ${requestUrl}...'`, this.LOG_CONTEXT);
            const token = yield this.tokenProvider.getAccessToken();
            const { data } = yield axios_1.default.get(requestUrl.href, {
                headers: { Authorization: `Bearer ${token}` },
                params: {
                    name: criteria === null || criteria === void 0 ? void 0 : criteria.name,
                    realm: criteria === null || criteria === void 0 ? void 0 : criteria.realm,
                },
            });
            const tenants = (_a = data === null || data === void 0 ? void 0 : data.results) === null || _a === void 0 ? void 0 : _a.map((r) => ({
                id: utils_1.AdspId.parse(r.id),
                name: r.name,
                realm: r.realm,
            }));
            return tenants;
        }));
        this.getTenants = () => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const tenants = yield this.retrieveTenants();
            return tenants;
        });
        this.getTenant = (tenantId) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            (0, utils_1.assertAdspId)(tenantId, `Provided ID does not represent a resource: ${tenantId}.`, 'resource');
            const cacheKey = `${tenantId}`;
            const tenant = tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenants, "f").get(cacheKey) || (yield this.retrieveTenant(tenantId));
            return tenant;
        });
        this.getTenantByName = (name) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!name) {
                return;
            }
            else {
                let tenantId = tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenantNames, "f")[name.toLowerCase()];
                if (!tenantId) {
                    // Query for the tenant if no existing record.
                    yield this.retrieveTenants({ name });
                    tenantId = tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenantNames, "f")[name.toLowerCase()];
                }
                return tenantId ? this.getTenant(tenantId) : null;
            }
        });
        this.getTenantByRealm = (realm) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!realm) {
                return;
            }
            else {
                let tenantId = tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenantRealms, "f")[realm.toLowerCase()];
                if (!tenantId) {
                    // Query for the tenant if no existing record.
                    yield this.retrieveTenants({ realm });
                    tenantId = tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenantRealms, "f")[realm.toLowerCase()];
                }
                return tenantId ? this.getTenant(tenantId) : null;
            }
        });
        if (preload) {
            // load into the cache.
            this.retrieveTenants().catch((err) => logger.error(`Encountered error during initialization of tenants. ${err}`));
        }
    }
    retrieveTenants(criteria) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const tenantServiceUrl = yield this.directory.getServiceUrl((0, utils_1.adspId) `urn:ads:platform:tenant-service:v2`);
            const tenantsUrl = new URL('v2/tenants', tenantServiceUrl);
            try {
                const tenants = yield retry((next, count) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    try {
                        return yield tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tryRetrieveTenants, "f").call(this, tenantsUrl, count, criteria);
                    }
                    catch (err) {
                        this.logger.debug(`Try ${count} failed with error. ${err}`, this.LOG_CONTEXT);
                        next(err);
                    }
                }));
                if (tenants) {
                    tenants.forEach((t) => {
                        const key = `${t.id}`;
                        tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenants, "f").set(key, t);
                        tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenantNames, "f")[t.name.toLowerCase()] = t.id;
                        tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenantRealms, "f")[t.realm.toLowerCase()] = t.id;
                        this.logger.debug(`Cached tenant '${key}' -> ${t.name} (${t.realm})`, this.LOG_CONTEXT);
                    });
                }
                this.logger.info(`Retrieved and cached tenant information.`, this.LOG_CONTEXT);
                return tenants;
            }
            catch (err) {
                this.logger.error(`Error encountered during retrieval of tenants. ${err}`, this.LOG_CONTEXT);
                throw err;
            }
        });
    }
    retrieveTenant(tenantId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Retrieving tenant information...'`, this.LOG_CONTEXT);
            const tenantServiceUrl = yield this.directory.getServiceUrl((0, utils_1.adspId) `urn:ads:platform:tenant-service:v2`);
            // Need the v2 prefix for a relative url of the form v2/tenants/... so the url resolves properly
            const tenantUrl = new URL(`v2${tenantId.resource}`, tenantServiceUrl);
            this.logger.debug(`Requesting tenant '${tenantId}' from ${tenantUrl}...'`, this.LOG_CONTEXT);
            try {
                const token = yield this.tokenProvider.getAccessToken();
                const { data } = yield axios_1.default.get(tenantUrl.href, {
                    headers: { Authorization: `Bearer ${token}` },
                });
                const tenant = data
                    ? Object.assign(Object.assign({}, data), { id: utils_1.AdspId.parse(data.id) }) : null;
                if (tenant) {
                    tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenants, "f").set(`${tenant.id}`, tenant);
                    tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenantNames, "f")[tenant.name.toLowerCase()] = tenant.id;
                    tslib_1.__classPrivateFieldGet(this, _TenantServiceImpl_tenantRealms, "f")[tenant.realm.toLowerCase()] = tenant.id;
                    this.logger.debug(`Cached tenant '${tenant.id}' -> ${tenant.name} (${tenant.realm})`, this.LOG_CONTEXT);
                }
                this.logger.debug(`Retrieved and cached tenant '${tenantId}'.`, this.LOG_CONTEXT);
                return tenant;
            }
            catch (err) {
                this.logger.error(`Error encountered during retrieval of tenant '${tenantId}'. ${err}`, this.LOG_CONTEXT);
                throw err;
            }
        });
    }
}
exports.TenantServiceImpl = TenantServiceImpl;
_TenantServiceImpl_tenants = new WeakMap(), _TenantServiceImpl_tenantNames = new WeakMap(), _TenantServiceImpl_tenantRealms = new WeakMap(), _TenantServiceImpl_tryRetrieveTenants = new WeakMap();
tslib_1.__decorate([
    (0, utils_1.LimitToOne)((propertyKey, criteria) => (criteria ? null : propertyKey)),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", Promise)
], TenantServiceImpl.prototype, "retrieveTenants", null);
tslib_1.__decorate([
    (0, utils_1.LimitToOne)((propertyKey, tenantId) => `${propertyKey}-${tenantId}`),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [utils_1.AdspId]),
    tslib_1.__metadata("design:returntype", Promise)
], TenantServiceImpl.prototype, "retrieveTenant", null);
