// @strapi/admin/dist/admin/admin/src/components/MainNav/NavBurgerMenu.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
require('react');
var designSystem = require('@strapi/design-system');
var styled = require('styled-components');
var theme = require('../../constants/theme.js');
var MainNavLinks = require('./MainNavLinks.js');
var NavUser = require('./NavUser.js');

const NavBurgerMenuWrapper = styled.styled(designSystem.Box)`
  position: fixed;
  top: calc(${theme.HEIGHT_TOP_NAVIGATION} + 1px);
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 3;
  background-color: ${({ theme })=>theme.colors.neutral0};
  transform: ${({ $isShown })=>$isShown ? 'translateY(0)' : 'translateY(-100%)'};
  transition: transform 0.2s ease-in-out;

  ${({ theme })=>theme.breakpoints.large} {
    display: none;
  }
`;
const NavBurgerMenu = ({ isShown, handleClickOnLink, listLinks })=>/*#__PURE__*/ jsxRuntime.jsx(NavBurgerMenuWrapper, {
        $isShown: isShown,
        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.ScrollArea, {
            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Box, {
                tag: "ul",
                paddingLeft: {
                    initial: 4,
                    medium: 6
                },
                paddingRight: {
                    initial: 4,
                    medium: 6
                },
                paddingTop: {
                    initial: 1,
                    medium: 3
                },
                paddingBottom: {
                    initial: 4,
                    medium: 6
                },
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(MainNavLinks.MainNavBurgerMenuLinks, {
                        listLinks: listLinks,
                        handleClickOnLink: handleClickOnLink
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                        paddingTop: 4,
                        tag: "li",
                        children: /*#__PURE__*/ jsxRuntime.jsx(NavUser.NavUser, {
                            showDisplayName: true
                        })
                    })
                ]
            })
        })
    });

exports.NavBurgerMenu = NavBurgerMenu;
