// @strapi/admin/dist/admin/admin/src/components/MainNav/NavBurgerMenu.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import 'react';
import { Box, ScrollArea } from '@strapi/design-system';
import { styled } from 'styled-components';
import { HEIGHT_TOP_NAVIGATION } from '../../constants/theme.mjs';
import { MainNavBurgerMenuLinks } from './MainNavLinks.mjs';
import { NavUser } from './NavUser.mjs';

const NavBurgerMenuWrapper = styled(Box)`
  position: fixed;
  top: calc(${HEIGHT_TOP_NAVIGATION} + 1px);
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 3;
  background-color: ${({ theme })=>theme.colors.neutral0};
  transform: ${({ $isShown })=>$isShown ? 'translateY(0)' : 'translateY(-100%)'};
  transition: transform 0.2s ease-in-out;

  ${({ theme })=>theme.breakpoints.large} {
    display: none;
  }
`;
const NavBurgerMenu = ({ isShown, handleClickOnLink, listLinks })=>/*#__PURE__*/ jsx(NavBurgerMenuWrapper, {
        $isShown: isShown,
        children: /*#__PURE__*/ jsx(ScrollArea, {
            children: /*#__PURE__*/ jsxs(Box, {
                tag: "ul",
                paddingLeft: {
                    initial: 4,
                    medium: 6
                },
                paddingRight: {
                    initial: 4,
                    medium: 6
                },
                paddingTop: {
                    initial: 1,
                    medium: 3
                },
                paddingBottom: {
                    initial: 4,
                    medium: 6
                },
                children: [
                    /*#__PURE__*/ jsx(MainNavBurgerMenuLinks, {
                        listLinks: listLinks,
                        handleClickOnLink: handleClickOnLink
                    }),
                    /*#__PURE__*/ jsx(Box, {
                        paddingTop: 4,
                        tag: "li",
                        children: /*#__PURE__*/ jsx(NavUser, {
                            showDisplayName: true
                        })
                    })
                ]
            })
        })
    });

export { NavBurgerMenu };
