// nx/src/plugins/js/project-graph/build-dependencies/explicit-package-json-dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildExplicitPackageJsonDependencies = buildExplicitPackageJsonDependencies;
const project_graph_1 = require("../../../../config/project-graph");
const file_utils_1 = require("../../../../project-graph/file-utils");
const project_graph_builder_1 = require("../../../../project-graph/project-graph-builder");
const json_1 = require("../../../../utils/json");
const path_1 = require("../../../../utils/path");
function buildExplicitPackageJsonDependencies(ctx, targetProjectLocator) {
    const res = [];
    // Build Set of valid package.json paths once for O(1) lookup
    // instead of O(n) find() per file
    const projectPackageJsonPaths = new Set(Object.values(ctx.projects).map((project) => (0, path_1.joinPathFragments)(project.root, 'package.json')));
    for (const source in ctx.filesToProcess.projectFileMap) {
        for (const f of Object.values(ctx.filesToProcess.projectFileMap[source])) {
            if (projectPackageJsonPaths.has(f.file)) {
                processPackageJson(source, f.file, ctx, targetProjectLocator, res);
            }
        }
    }
    return res;
}
function processPackageJson(sourceProject, packageJsonPath, ctx, targetProjectLocator, collectedDeps) {
    try {
        const deps = readDeps((0, json_1.parseJson)((0, file_utils_1.defaultFileRead)(packageJsonPath)));
        for (const [packageName, packageVersion] of Object.entries(deps)) {
            const localProject = targetProjectLocator.findDependencyInWorkspaceProjects(packageJsonPath, packageName, packageVersion);
            if (localProject) {
                // package.json refers to another project in the monorepo
                const dependency = {
                    source: sourceProject,
                    target: localProject,
                    sourceFile: packageJsonPath,
                    type: project_graph_1.DependencyType.static,
                };
                (0, project_graph_builder_1.validateDependency)(dependency, ctx);
                collectedDeps.push(dependency);
                continue;
            }
            const externalNodeName = targetProjectLocator.findNpmProjectFromImport(packageName, packageJsonPath);
            if (!externalNodeName) {
                continue;
            }
            const dependency = {
                source: sourceProject,
                target: externalNodeName,
                sourceFile: packageJsonPath,
                type: project_graph_1.DependencyType.static,
            };
            (0, project_graph_builder_1.validateDependency)(dependency, ctx);
            collectedDeps.push(dependency);
        }
    }
    catch (e) {
        if (process.env.NX_VERBOSE_LOGGING === 'true') {
            console.error(e);
        }
    }
}
function readDeps(packageJson) {
    const deps = {};
    /**
     * We process dependencies in a rough order of increasing importance such that if a dependency is listed in multiple
     * sections, the version listed under the "most important" one wins, with production dependencies being the most important.
     */
    const depType = [
        'optionalDependencies',
        'peerDependencies',
        'devDependencies',
        'dependencies',
    ];
    for (const type of depType) {
        for (const [depName, depVersion] of Object.entries(packageJson[type] || {})) {
            deps[depName] = depVersion;
        }
    }
    return deps;
}
