// @abgov/adsp-service-sdk/src/configuration/configurationUpdateHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleConfigurationUpdates = void 0;
const tslib_1 = require("tslib");
const socket_io_client_1 = require("socket.io-client");
const utils_1 = require("../utils");
const LOG_CONTEXT = { context: 'handleConfigurationUpdates' };
const handleConfigurationUpdates = (logger, directory, tokenProvider, configurationService) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const pushServiceUrl = yield directory.getServiceUrl((0, utils_1.adspId) `urn:ads:platform:push-service`);
    const streamUrl = new URL('', pushServiceUrl);
    const socket = (0, socket_io_client_1.io)(streamUrl.href, {
        autoConnect: true,
        reconnection: true,
        query: {
            stream: 'configuration-updates',
        },
        transports: ['websocket'],
        withCredentials: true,
        auth: (cb) => tslib_1.__awaiter(void 0, void 0, void 0, function* () { return cb({ token: yield tokenProvider.getAccessToken() }); }),
    });
    socket.on('connect', () => {
        logger.info('Connected for configuration updates...', LOG_CONTEXT);
    });
    socket.on('connect_error', (err) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        logger.error(`Connect to configuration updates failed with error: ${err}`, LOG_CONTEXT);
    }));
    socket.on('disconnect', (reason) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        logger.debug(`Disconnected from configuration updates due to reason: ${reason}`);
    }));
    const invalidateCached = (e) => {
        const tenantId = utils_1.AdspId.parse(e.tenantId);
        const { namespace, name } = e.payload;
        logger.debug(`Received configuration updated stream item ${namespace}:${name} and invalidating cache of tenant ${tenantId}`, Object.assign(Object.assign({}, LOG_CONTEXT), { tenantId: e.tenantId }));
        configurationService.clearCached(tenantId, namespace, name);
    };
    socket.on('configuration-service:configuration-updated', invalidateCached);
    socket.on('configuration-service:active-revision-set', invalidateCached);
});
exports.handleConfigurationUpdates = handleConfigurationUpdates;
