// @strapi/admin/dist/admin/admin/src/services/homepage.mjs
import { adminApi } from './api.mjs';

const homepageService = adminApi.enhanceEndpoints({
    addTagTypes: [
        'CountDocuments'
    ]
}).injectEndpoints({
    endpoints: (builder)=>({
            getKeyStatistics: builder.query({
                query: ()=>'/admin/homepage/key-statistics',
                transformResponse: (response)=>response.data,
                providesTags: (_, _err)=>[
                        'HomepageKeyStatistics'
                    ]
            }),
            getCountDocuments: builder.query({
                query: ()=>'/content-manager/homepage/count-documents',
                transformResponse: (response)=>response.data,
                providesTags: (_, _err)=>[
                        'CountDocuments'
                    ]
            }),
            getHomepageLayout: builder.query({
                query: ()=>'/admin/homepage/layout',
                transformResponse: (r)=>r.data
            }),
            updateHomepageLayout: builder.mutation({
                query: (body)=>({
                        url: '/admin/homepage/layout',
                        method: 'PUT',
                        body
                    }),
                transformResponse: (r)=>r.data
            })
        })
});
const { useGetKeyStatisticsQuery, useGetCountDocumentsQuery, useGetHomepageLayoutQuery, useUpdateHomepageLayoutMutation } = homepageService;

export { useGetCountDocumentsQuery, useGetHomepageLayoutQuery, useGetKeyStatisticsQuery, useUpdateHomepageLayoutMutation };
