// @strapi/admin/dist/admin/admin/src/services/homepage.js
'use strict';

var api = require('./api.js');

const homepageService = api.adminApi.enhanceEndpoints({
    addTagTypes: [
        'CountDocuments'
    ]
}).injectEndpoints({
    endpoints: (builder)=>({
            getKeyStatistics: builder.query({
                query: ()=>'/admin/homepage/key-statistics',
                transformResponse: (response)=>response.data,
                providesTags: (_, _err)=>[
                        'HomepageKeyStatistics'
                    ]
            }),
            getCountDocuments: builder.query({
                query: ()=>'/content-manager/homepage/count-documents',
                transformResponse: (response)=>response.data,
                providesTags: (_, _err)=>[
                        'CountDocuments'
                    ]
            }),
            getHomepageLayout: builder.query({
                query: ()=>'/admin/homepage/layout',
                transformResponse: (r)=>r.data
            }),
            updateHomepageLayout: builder.mutation({
                query: (body)=>({
                        url: '/admin/homepage/layout',
                        method: 'PUT',
                        body
                    }),
                transformResponse: (r)=>r.data
            })
        })
});
const { useGetKeyStatisticsQuery, useGetCountDocumentsQuery, useGetHomepageLayoutQuery, useUpdateHomepageLayoutMutation } = homepageService;

exports.useGetCountDocumentsQuery = useGetCountDocumentsQuery;
exports.useGetHomepageLayoutQuery = useGetHomepageLayoutQuery;
exports.useGetKeyStatisticsQuery = useGetKeyStatisticsQuery;
exports.useUpdateHomepageLayoutMutation = useUpdateHomepageLayoutMutation;
