// @abgov/adsp-service-sdk/src/registration/registration.js
"use strict";
var _ServiceRegistrarImpl_tryUpdateConfiguration;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceRegistrarImpl = void 0;
const tslib_1 = require("tslib");
const axios_1 = require("axios");
// eslint-disable-next-line @typescript-eslint/no-var-requires
const retry = require('promise-retry');
const utils_1 = require("../utils");
class ServiceRegistrarImpl {
    constructor(logger, directory, tokenProvider) {
        this.logger = logger;
        this.directory = directory;
        this.tokenProvider = tokenProvider;
        this.LOG_CONTEXT = { context: 'ServiceRegistration' };
        _ServiceRegistrarImpl_tryUpdateConfiguration.set(this, (configurationServiceUrl, count, serviceId, update) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Try ${count}: updating configuration for service ${serviceId}...`, this.LOG_CONTEXT);
            const configurationUrl = new URL(`v2/configuration/${serviceId.namespace}/${serviceId.service}`, configurationServiceUrl);
            const token = yield this.tokenProvider.getAccessToken();
            yield axios_1.default.patch(configurationUrl.href, {
                operation: 'UPDATE',
                update,
            }, { headers: { Authorization: `Bearer ${token}` } });
        }));
    }
    register(registration) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (registration.configurationSchema) {
                const namespace = registration.serviceId.namespace;
                const name = registration.serviceId.service;
                const update = {
                    [`${namespace}:${name}`]: {
                        configurationSchema: registration.configurationSchema,
                    },
                };
                yield this.updateConfiguration((0, utils_1.adspId) `urn:ads:platform:configuration-service`, update);
            }
            if (registration.configuration) {
                const namespace = registration.serviceId.namespace;
                const name = registration.serviceId.service;
                // If useNamespace is enabled, then the service is the configuration namespace and no name is specified.
                // Used in cases where the service can have a lot of configuration, and cannot store all of it in a single document.
                const key = registration.configuration.useNamespace ? name : `${namespace}:${name}`;
                const update = {
                    [key]: {
                        configurationSchema: registration.configuration.schema,
                        description: registration.configuration.description,
                    },
                };
                yield this.updateConfiguration((0, utils_1.adspId) `urn:ads:platform:configuration-service`, update);
            }
            if (registration.roles) {
                const update = {
                    [registration.serviceId.toString()]: {
                        roles: registration.roles.map((r) => (r.role ? r : { role: r, description: '' })) || [],
                    },
                };
                yield this.updateConfiguration((0, utils_1.adspId) `urn:ads:platform:tenant-service`, update);
            }
            if (registration.events) {
                const namespace = registration.serviceId.service;
                const update = {
                    [namespace]: {
                        name: namespace,
                        definitions: registration.events.reduce((defs, def) => (Object.assign(Object.assign({}, defs), { [def.name]: {
                                name: def.name,
                                description: def.description,
                                payloadSchema: def.payloadSchema,
                                interval: def.interval,
                                log: def.log,
                            } })), {}),
                    },
                };
                yield this.updateConfiguration((0, utils_1.adspId) `urn:ads:platform:event-service`, update);
            }
            if (registration.eventStreams) {
                const update = registration.eventStreams.reduce((streams, stream) => (Object.assign(Object.assign({}, streams), { [stream.id]: {
                        id: stream.id,
                        name: stream.name,
                        description: stream.description,
                        events: stream.events,
                        subscriberRoles: stream.subscriberRoles,
                        publicSubscribe: stream.publicSubscribe,
                    } })), {});
                yield this.updateConfiguration((0, utils_1.adspId) `urn:ads:platform:push-service`, update);
            }
            if (registration.fileTypes) {
                const update = registration.fileTypes.reduce((types, type) => (Object.assign(Object.assign({}, types), { [type.id]: {
                        id: type.id,
                        name: type.name,
                        anonymousRead: type.anonymousRead,
                        securityClassification: type.securityClassification,
                        readRoles: type.readRoles,
                        updateRoles: type.updateRoles,
                        rules: type === null || type === void 0 ? void 0 : type.rules,
                    } })), {});
                yield this.updateConfiguration((0, utils_1.adspId) `urn:ads:platform:file-service`, update);
            }
            if (registration.notifications) {
                const update = registration.notifications.reduce((types, type) => (Object.assign(Object.assign({}, types), { [type.name]: Object.assign(Object.assign({}, type), { id: type.name, name: type.displayName || type.name, events: type.events }) })), {});
                yield this.updateConfiguration((0, utils_1.adspId) `urn:ads:platform:notification-service`, update);
            }
            if (registration.values) {
                const namespace = registration.serviceId.service;
                const update = {
                    [namespace]: {
                        name: namespace,
                        definitions: registration.values.reduce((defs, def) => (Object.assign(Object.assign({}, defs), { [def.id]: {
                                name: def.id,
                                displayName: def.name,
                                description: def.description,
                                jsonSchema: def.jsonSchema,
                            } })), {}),
                    },
                };
                yield this.updateConfiguration((0, utils_1.adspId) `urn:ads:platform:value-service`, update);
            }
            if (registration.serviceConfigurations) {
                for (const { serviceId, configuration } of registration.serviceConfigurations) {
                    yield this.updateConfiguration(serviceId, configuration);
                }
            }
        });
    }
    updateConfiguration(serviceId, update) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const configurationServiceId = (0, utils_1.adspId) `urn:ads:platform:configuration-service:v2`;
            const serviceUrl = yield this.directory.getServiceUrl(configurationServiceId);
            try {
                yield retry((next, count) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    try {
                        yield tslib_1.__classPrivateFieldGet(this, _ServiceRegistrarImpl_tryUpdateConfiguration, "f").call(this, serviceUrl, count, serviceId, update);
                    }
                    catch (err) {
                        this.logger.debug(`Try ${count} failed with error. ${err}`, this.LOG_CONTEXT);
                        next(err);
                    }
                }));
                this.logger.info(`Updated registration configuration of ${serviceId}.`, this.LOG_CONTEXT);
            }
            catch (err) {
                this.logger.error(`Error encountered registering configuration. ${err}`);
                throw err;
            }
        });
    }
}
exports.ServiceRegistrarImpl = ServiceRegistrarImpl;
_ServiceRegistrarImpl_tryUpdateConfiguration = new WeakMap();
