// @abgov/adsp-service-sdk/src/event/eventService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventServiceImpl = void 0;
const tslib_1 = require("tslib");
const axios_1 = require("axios");
const utils_1 = require("../utils");
class EventServiceImpl {
    constructor(isCore, logger, directory, tokenProvider, serviceId, events) {
        this.isCore = isCore;
        this.logger = logger;
        this.directory = directory;
        this.tokenProvider = tokenProvider;
        this.LOG_CONTEXT = { context: 'EventService' };
        this.namespace = serviceId.service;
        this.definitions = (events === null || events === void 0 ? void 0 : events.map((e) => e.name)) || [];
    }
    send(_a) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            var _b, _c;
            var { tenantId } = _a, event = tslib_1.__rest(_a, ["tenantId"]);
            if (!this.definitions.includes(event.name)) {
                throw new Error(`Event ${this.namespace}:${event.name} is not recognized; only registered events can be sent.`);
            }
            const serviceUrl = yield this.directory.getServiceUrl((0, utils_1.adspId) `urn:ads:platform:event-service:v1`);
            const sendUrl = new URL('v1/events', serviceUrl);
            try {
                const token = yield this.tokenProvider.getAccessToken();
                this.logger.debug(`Sending event ${this.namespace}:${event.name} to: ${sendUrl}...`, Object.assign(Object.assign({}, this.LOG_CONTEXT), { tenant: tenantId === null || tenantId === void 0 ? void 0 : tenantId.toString() }));
                yield axios_1.default.post(sendUrl.href, this.isCore
                    ? Object.assign(Object.assign({}, event), { namespace: this.namespace, tenantId: `${tenantId}` }) : Object.assign(Object.assign({}, event), { namespace: this.namespace }), { headers: { Authorization: `Bearer ${token}` } });
                this.logger.info(`Sent domain event ${this.namespace}:${event.name}.`, Object.assign(Object.assign({}, this.LOG_CONTEXT), { tenant: tenantId === null || tenantId === void 0 ? void 0 : tenantId.toString() }));
            }
            catch (err) {
                this.logger.error(`Error encountered on sending of event ${this.namespace}:${event.name}. ${err}`, Object.assign(Object.assign({}, this.LOG_CONTEXT), { tenant: tenantId === null || tenantId === void 0 ? void 0 : tenantId.toString() }));
                if (axios_1.default.isAxiosError(err) && ((_c = (_b = err.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.errorMessage)) {
                    this.logger.debug(`Event send failed with request status ${err.response.status} and error and message: ${err.response.data.errorMessage}`);
                }
            }
        });
    }
}
exports.EventServiceImpl = EventServiceImpl;
