// @strapi/content-manager/dist/admin/pages/EditView/components/DocumentActions.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var Icons = require('@strapi/icons');
var mapValues = require('lodash/fp/mapValues');
var get = require('lodash/get');
var merge = require('lodash/merge');
var set = require('lodash/set');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var attributes = require('../../../constants/attributes.js');
var collections = require('../../../constants/collections.js');
var DocumentRBAC = require('../../../features/DocumentRBAC.js');
var useDocument = require('../../../hooks/useDocument.js');
var useDocumentActions = require('../../../hooks/useDocumentActions.js');
var useDocumentContext = require('../../../hooks/useDocumentContext.js');
var Preview = require('../../../preview/pages/Preview.js');
var router = require('../../../router.js');
var documents = require('../../../services/documents.js');
var api = require('../../../utils/api.js');
var translations = require('../../../utils/translations.js');
var data = require('../utils/data.js');
var RelationModal = require('./FormInputs/Relations/RelationModal.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const connectRelationToParent = (parentDataToUpdate, fieldToConnect, data, fieldToConnectUID)=>{
    /*
   * Check if the fieldToConnect is already present in the parentDataToUpdate.
   * This happens in particular when in the parentDocument you have created
   * a new component without saving.
   */ const isFieldPresent = !!get(parentDataToUpdate, fieldToConnect);
    const fieldToConnectPath = isFieldPresent ? fieldToConnect : fieldToConnect.split('.').slice(0, -1).join('.');
    const fieldToConnectValue = isFieldPresent ? {
        connect: [
            {
                id: data.documentId,
                documentId: data.documentId,
                locale: data.locale
            }
        ]
    } : {
        [fieldToConnect.split('.').pop()]: {
            connect: [
                {
                    id: data.documentId,
                    documentId: data.documentId,
                    locale: data.locale
                }
            ],
            disconnect: []
        },
        // In case the object was not present you need to pass the componentUID of the parent document
        __component: fieldToConnectUID
    };
    const objectToConnect = set({}, fieldToConnectPath, fieldToConnectValue);
    return merge(parentDataToUpdate, objectToConnect);
};
const DocumentActions = ({ actions })=>{
    const { formatMessage } = reactIntl.useIntl();
    const [primaryAction, secondaryAction, ...restActions] = actions.filter((action)=>{
        if (action.position === undefined) {
            return true;
        }
        const positions = Array.isArray(action.position) ? action.position : [
            action.position
        ];
        return positions.includes('panel');
    });
    if (!primaryAction) {
        return null;
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
        direction: "column",
        gap: 2,
        alignItems: "stretch",
        width: "100%",
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.tours.contentManager.Publish, {
                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                    gap: 2,
                    children: [
                        primaryAction.label === 'Publish' ? /*#__PURE__*/ jsxRuntime.jsx(DocumentActionButton, {
                            ...primaryAction,
                            variant: primaryAction.variant || 'default'
                        }) : /*#__PURE__*/ jsxRuntime.jsx(DocumentActionButton, {
                            ...primaryAction,
                            variant: primaryAction.variant || 'default'
                        }),
                        restActions.length > 0 ? /*#__PURE__*/ jsxRuntime.jsx(DocumentActionsMenu, {
                            actions: restActions,
                            label: formatMessage({
                                id: 'content-manager.containers.edit.panels.default.more-actions',
                                defaultMessage: 'More document actions'
                            })
                        }) : null
                    ]
                })
            }),
            secondaryAction ? secondaryAction.label === 'Publish' ? /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.tours.contentManager.Publish, {
                children: /*#__PURE__*/ jsxRuntime.jsx(DocumentActionButton, {
                    ...secondaryAction,
                    variant: secondaryAction.variant || 'secondary'
                })
            }) : /*#__PURE__*/ jsxRuntime.jsx(DocumentActionButton, {
                ...secondaryAction,
                variant: secondaryAction.variant || 'secondary'
            }) : null
        ]
    });
};
const DocumentActionButton = (action)=>{
    const [dialogId, setDialogId] = React__namespace.useState(null);
    const { toggleNotification } = strapiAdmin.useNotification();
    const handleClick = (action)=>async (e)=>{
            const { onClick = ()=>false, dialog, id } = action;
            const muteDialog = await onClick(e);
            if (dialog && !muteDialog) {
                switch(dialog.type){
                    case 'notification':
                        toggleNotification({
                            title: dialog.title,
                            message: dialog.content,
                            type: dialog.status,
                            timeout: dialog.timeout,
                            onClose: dialog.onClose
                        });
                        break;
                    case 'dialog':
                    case 'modal':
                        e.preventDefault();
                        setDialogId(id);
                }
            }
        };
    const handleClose = ()=>{
        setDialogId(null);
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                flex: "auto",
                startIcon: action.icon,
                disabled: action.disabled,
                onClick: handleClick(action),
                justifyContent: "center",
                variant: action.variant || 'default',
                paddingTop: "7px",
                paddingBottom: "7px",
                loading: action.loading,
                children: action.label
            }),
            action.dialog?.type === 'dialog' ? /*#__PURE__*/ jsxRuntime.jsx(DocumentActionConfirmDialog, {
                ...action.dialog,
                variant: action.dialog?.variant ?? action.variant,
                isOpen: dialogId === action.id,
                onClose: handleClose
            }) : null,
            action.dialog?.type === 'modal' ? /*#__PURE__*/ jsxRuntime.jsx(DocumentActionModal, {
                ...action.dialog,
                onModalClose: handleClose,
                isOpen: dialogId === action.id
            }) : null
        ]
    });
};
const DocumentActionsMenu = ({ actions, children, label, variant = 'tertiary' })=>{
    const [isOpen, setIsOpen] = React__namespace.useState(false);
    const [dialogId, setDialogId] = React__namespace.useState(null);
    const { formatMessage } = reactIntl.useIntl();
    const { toggleNotification } = strapiAdmin.useNotification();
    const isDisabled = actions.every((action)=>action.disabled) || actions.length === 0;
    const handleClick = (action)=>async (e)=>{
            const { onClick = ()=>false, dialog, id } = action;
            const muteDialog = await onClick(e);
            if (dialog && !muteDialog) {
                switch(dialog.type){
                    case 'notification':
                        toggleNotification({
                            title: dialog.title,
                            message: dialog.content,
                            type: dialog.status,
                            timeout: dialog.timeout,
                            onClose: dialog.onClose
                        });
                        break;
                    case 'dialog':
                    case 'modal':
                        setDialogId(id);
                }
            }
        };
    const handleClose = ()=>{
        setDialogId(null);
        setIsOpen(false);
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Menu.Root, {
        open: isOpen,
        onOpenChange: setIsOpen,
        children: [
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Menu.Trigger, {
                disabled: isDisabled,
                size: "S",
                endIcon: null,
                paddingTop: "4px",
                paddingLeft: "7px",
                paddingRight: "7px",
                variant: variant,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(Icons.More, {
                        "aria-hidden": true,
                        focusable: false
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.VisuallyHidden, {
                        tag: "span",
                        children: label || formatMessage({
                            id: 'content-manager.containers.edit.panels.default.more-actions',
                            defaultMessage: 'More document actions'
                        })
                    })
                ]
            }),
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Menu.Content, {
                maxHeight: undefined,
                popoverPlacement: "bottom-end",
                children: [
                    actions.map((action)=>{
                        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Menu.Item, {
                            disabled: action.disabled,
                            /* @ts-expect-error – TODO: this is an error in the DS where it is most likely a synthetic event, not regular. */ onSelect: handleClick(action),
                            display: "block",
                            variant: action.variant === 'danger' ? action.variant : 'default',
                            startIcon: action.icon,
                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                                justifyContent: "space-between",
                                gap: 4,
                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                                    gap: 2,
                                    tag: "span",
                                    children: action.label
                                })
                            })
                        }, action.id);
                    }),
                    children
                ]
            }),
            actions.map((action)=>{
                return /*#__PURE__*/ jsxRuntime.jsxs(React__namespace.Fragment, {
                    children: [
                        action.dialog?.type === 'dialog' ? /*#__PURE__*/ jsxRuntime.jsx(DocumentActionConfirmDialog, {
                            ...action.dialog,
                            variant: action.variant,
                            isOpen: dialogId === action.id,
                            onClose: handleClose
                        }) : null,
                        action.dialog?.type === 'modal' ? /*#__PURE__*/ jsxRuntime.jsx(DocumentActionModal, {
                            ...action.dialog,
                            onModalClose: handleClose,
                            isOpen: dialogId === action.id
                        }) : null
                    ]
                }, action.id);
            })
        ]
    });
};
const DocumentActionConfirmDialog = ({ onClose, onCancel, onConfirm, title, content, isOpen, variant = 'secondary', loading })=>{
    const { formatMessage } = reactIntl.useIntl();
    const handleClose = async ()=>{
        if (onCancel) {
            await onCancel();
        }
        onClose();
    };
    const handleConfirm = async ()=>{
        if (onConfirm) {
            await onConfirm();
        }
        onClose();
    };
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Root, {
        open: isOpen,
        onOpenChange: handleClose,
        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Dialog.Content, {
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Header, {
                    children: title
                }),
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Body, {
                    children: content
                }),
                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Dialog.Footer, {
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Cancel, {
                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                variant: "tertiary",
                                fullWidth: true,
                                children: formatMessage({
                                    id: 'app.components.Button.cancel',
                                    defaultMessage: 'Cancel'
                                })
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                            onClick: handleConfirm,
                            variant: variant,
                            fullWidth: true,
                            loading: loading,
                            children: formatMessage({
                                id: 'app.components.Button.confirm',
                                defaultMessage: 'Confirm'
                            })
                        })
                    ]
                })
            ]
        })
    });
};
const DocumentActionModal = ({ isOpen, title, onClose, footer: Footer, content: Content, onModalClose })=>{
    const handleClose = ()=>{
        if (onClose) {
            onClose();
        }
        onModalClose();
    };
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Root, {
        open: isOpen,
        onOpenChange: handleClose,
        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Modal.Content, {
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Header, {
                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Title, {
                        children: title
                    })
                }),
                typeof Content === 'function' ? /*#__PURE__*/ jsxRuntime.jsx(Content, {
                    onClose: handleClose
                }) : /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Body, {
                    children: Content
                }),
                typeof Footer === 'function' ? /*#__PURE__*/ jsxRuntime.jsx(Footer, {
                    onClose: handleClose
                }) : Footer
            ]
        })
    });
};
const transformData = (data)=>{
    if (Array.isArray(data)) {
        return data.map(transformData);
    }
    if (typeof data === 'object' && data !== null) {
        if ('apiData' in data) {
            return data.apiData;
        }
        return mapValues(transformData)(data);
    }
    return data;
};
/* -------------------------------------------------------------------------------------------------
 * DocumentActionComponents
 * -----------------------------------------------------------------------------------------------*/ const PublishAction = ({ activeTab, documentId, model, collectionType, meta, document })=>{
    const { currentDocument: { schema } } = useDocumentContext.useDocumentContext('PublishAction');
    const navigate = reactRouterDom.useNavigate();
    const { toggleNotification } = strapiAdmin.useNotification();
    const { _unstableFormatValidationErrors: formatValidationErrors } = strapiAdmin.useAPIErrorHandler();
    const isListView = reactRouterDom.useMatch(router.LIST_PATH) !== null;
    const isCloning = reactRouterDom.useMatch(router.CLONE_PATH) !== null;
    const { id } = reactRouterDom.useParams();
    const { formatMessage } = reactIntl.useIntl();
    const { canPublish, canReadFields } = DocumentRBAC.useDocumentRBAC('PublishAction', ({ canPublish, canReadFields })=>({
            canPublish,
            canReadFields
        }));
    const { publish, isLoading } = useDocumentActions.useDocumentActions();
    const onPreview = Preview.usePreviewContext('UpdateAction', (state)=>state.onPreview, false);
    const [countDraftRelations, { isLoading: isLoadingDraftRelations, isError: isErrorDraftRelations }] = documents.useGetDraftRelationCountQuery();
    const [localCountOfDraftRelations, setLocalCountOfDraftRelations] = React__namespace.useState(0);
    const [serverCountOfDraftRelations, setServerCountOfDraftRelations] = React__namespace.useState(0);
    const [{ rawQuery }] = strapiAdmin.useQueryParams();
    const modified = strapiAdmin.useForm('PublishAction', ({ modified })=>modified);
    const setSubmitting = strapiAdmin.useForm('PublishAction', ({ setSubmitting })=>setSubmitting);
    const isSubmitting = strapiAdmin.useForm('PublishAction', ({ isSubmitting })=>isSubmitting);
    const validate = strapiAdmin.useForm('PublishAction', (state)=>state.validate);
    const setErrors = strapiAdmin.useForm('PublishAction', (state)=>state.setErrors);
    const formValues = strapiAdmin.useForm('PublishAction', ({ values })=>values);
    const resetForm = strapiAdmin.useForm('PublishAction', ({ resetForm })=>resetForm);
    const { currentDocument: { components } } = useDocumentContext.useDocumentContext('PublishAction');
    // need to discriminate if the publish is coming from a relation modal or in the edit view
    const relationContext = RelationModal.useRelationModal('PublishAction', ()=>true, false);
    const fromRelationModal = relationContext != undefined;
    const dispatch = RelationModal.useRelationModal('PublishAction', (state)=>state.dispatch);
    const fieldToConnect = RelationModal.useRelationModal('PublishAction', (state)=>state.state.fieldToConnect, false);
    const fieldToConnectUID = RelationModal.useRelationModal('PublishAction', (state)=>state.state.fieldToConnectUID, false);
    const documentHistory = RelationModal.useRelationModal('PublishAction', (state)=>state.state.documentHistory, false);
    const rootDocumentMeta = RelationModal.useRelationModal('PublishAction', (state)=>state.rootDocumentMeta);
    const dispatchGuidedTour = strapiAdmin.useGuidedTour('PublishAction', (s)=>s.dispatch);
    const { currentDocumentMeta } = useDocumentContext.useDocumentContext('PublishAction');
    const [updateDocumentMutation] = documents.useUpdateDocumentMutation();
    const { _unstableFormatAPIError: formatAPIError } = strapiAdmin.useAPIErrorHandler();
    const idToPublish = currentDocumentMeta.documentId || id;
    React__namespace.useEffect(()=>{
        if (isErrorDraftRelations) {
            toggleNotification({
                type: 'danger',
                message: formatMessage({
                    id: translations.getTranslation('error.records.fetch-draft-relatons'),
                    defaultMessage: 'An error occurred while fetching draft relations on this document.'
                })
            });
        }
    }, [
        isErrorDraftRelations,
        toggleNotification,
        formatMessage
    ]);
    React__namespace.useEffect(()=>{
        const localDraftRelations = new Set();
        /**
     * Extracts draft relations from the provided data object.
     * It checks for a connect array of relations.
     * If a relation has a status of 'draft', its id is added to the localDraftRelations set.
     */ const extractDraftRelations = (data)=>{
            const relations = data.connect || [];
            relations.forEach((relation)=>{
                if (relation.status === 'draft') {
                    localDraftRelations.add(relation.id);
                }
            });
        };
        /**
     * Recursively traverses the provided data object to extract draft relations from arrays within 'connect' keys.
     * If the data is an object, it looks for 'connect' keys to pass their array values to extractDraftRelations.
     * It recursively calls itself for any non-null objects it contains.
     */ const traverseAndExtract = (data)=>{
            Object.entries(data).forEach(([key, value])=>{
                if (key === 'connect' && Array.isArray(value)) {
                    extractDraftRelations({
                        connect: value
                    });
                } else if (typeof value === 'object' && value !== null) {
                    traverseAndExtract(value);
                }
            });
        };
        if (!documentId || modified) {
            traverseAndExtract(formValues);
            setLocalCountOfDraftRelations(localDraftRelations.size);
        }
    }, [
        documentId,
        modified,
        formValues,
        setLocalCountOfDraftRelations
    ]);
    React__namespace.useEffect(()=>{
        if (!document || !document.documentId || isListView) {
            return;
        }
        const fetchDraftRelationsCount = async ()=>{
            const { data, error } = await countDraftRelations({
                collectionType,
                model,
                documentId,
                params: currentDocumentMeta.params
            });
            if (error) {
                throw error;
            }
            if (data) {
                setServerCountOfDraftRelations(data.data);
            }
        };
        fetchDraftRelationsCount();
    }, [
        isListView,
        document,
        documentId,
        countDraftRelations,
        collectionType,
        model,
        currentDocumentMeta.params
    ]);
    const parentDocumentMetaToUpdate = documentHistory?.at(-2) ?? rootDocumentMeta;
    const parentDocumentData = useDocument.useDocument({
        documentId: parentDocumentMetaToUpdate?.documentId,
        model: parentDocumentMetaToUpdate?.model,
        collectionType: parentDocumentMetaToUpdate?.collectionType,
        params: parentDocumentMetaToUpdate?.params
    }, {
        skip: !parentDocumentMetaToUpdate
    });
    const { getInitialFormValues } = useDocument.useDoc();
    const isDocumentPublished = (document?.[attributes.PUBLISHED_AT_ATTRIBUTE_NAME] || meta?.availableStatus.some((doc)=>doc[attributes.PUBLISHED_AT_ATTRIBUTE_NAME] !== null)) && document?.status !== 'modified';
    if (!schema?.options?.draftAndPublish) {
        return null;
    }
    const performPublish = async ()=>{
        setSubmitting(true);
        try {
            const { errors } = await validate(true, {
                status: 'published'
            });
            if (errors) {
                const hasUnreadableRequiredField = Object.keys(schema.attributes).some((fieldName)=>{
                    const attribute = schema.attributes[fieldName];
                    return attribute?.required && !(canReadFields ?? []).includes(fieldName);
                });
                if (hasUnreadableRequiredField) {
                    toggleNotification({
                        type: 'danger',
                        message: formatMessage({
                            id: 'content-manager.validation.error.unreadable-required-field',
                            defaultMessage: 'Your current permissions prevent access to certain required fields. Please request access from an administrator to proceed.'
                        })
                    });
                } else {
                    toggleNotification({
                        type: 'danger',
                        message: formatMessage({
                            id: 'content-manager.validation.error',
                            defaultMessage: 'There are validation errors in your document. Please fix them before saving.'
                        })
                    });
                }
                return;
            }
            const { data: data$1 } = data.handleInvisibleAttributes(transformData(formValues), {
                schema,
                components
            });
            const res = await publish({
                collectionType,
                model,
                documentId,
                params: currentDocumentMeta.params
            }, data$1);
            // Reset form if successful
            if ('data' in res) {
                resetForm();
                dispatchGuidedTour({
                    type: 'set_completed_actions',
                    payload: [
                        strapiAdmin.GUIDED_TOUR_REQUIRED_ACTIONS.contentManager.createContent
                    ]
                });
            }
            if ('data' in res && collectionType !== collections.SINGLE_TYPES) {
                /**
         * TODO: refactor the router so we can just do `../${res.data.documentId}` instead of this.
         */ if (idToPublish === 'create' && !fromRelationModal) {
                    navigate({
                        pathname: `../${collectionType}/${model}/${res.data.documentId}`,
                        search: rawQuery
                    });
                } else if (fromRelationModal) {
                    const newRelation = {
                        documentId: res.data.documentId,
                        collectionType,
                        model,
                        params: currentDocumentMeta.params
                    };
                    /*
           * Update, if needed, the parent relation with the newly published document.
           * Check if in history we have the parent relation otherwise use the
           * rootDocument
           */ if (fieldToConnect && documentHistory && (parentDocumentMetaToUpdate.documentId || parentDocumentMetaToUpdate.collectionType === collections.SINGLE_TYPES)) {
                        const parentDataToUpdate = parentDocumentMetaToUpdate.collectionType === collections.SINGLE_TYPES ? getInitialFormValues() : parentDocumentData.getInitialFormValues();
                        const metaDocumentToUpdate = documentHistory.at(-2) ?? rootDocumentMeta;
                        const dataToUpdate = connectRelationToParent(parentDataToUpdate, fieldToConnect, res.data, fieldToConnectUID);
                        try {
                            const updateRes = await updateDocumentMutation({
                                collectionType: metaDocumentToUpdate.collectionType,
                                model: metaDocumentToUpdate.model,
                                documentId: metaDocumentToUpdate.collectionType !== collections.SINGLE_TYPES ? metaDocumentToUpdate.documentId : undefined,
                                params: metaDocumentToUpdate.params,
                                data: dataToUpdate
                            });
                            if ('error' in updateRes) {
                                toggleNotification({
                                    type: 'danger',
                                    message: formatAPIError(updateRes.error)
                                });
                                return;
                            }
                        } catch (err) {
                            toggleNotification({
                                type: 'danger',
                                message: formatMessage({
                                    id: 'notification.error',
                                    defaultMessage: 'An error occurred'
                                })
                            });
                            throw err;
                        }
                    }
                    dispatch({
                        type: 'GO_TO_CREATED_RELATION',
                        payload: {
                            document: newRelation,
                            shouldBypassConfirmation: true
                        }
                    });
                }
            } else if ('error' in res && api.isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                setErrors(formatValidationErrors(res.error));
            }
        } finally{
            setSubmitting(false);
            if (onPreview) {
                onPreview();
            }
        }
    };
    const totalDraftRelations = localCountOfDraftRelations + serverCountOfDraftRelations;
    // TODO skipping this for now as there is a bug with the draft relation count that will be worked on separately
    // see RFC "Count draft relations" in Notion
    const enableDraftRelationsCount = false;
    const hasDraftRelations = enableDraftRelationsCount;
    return {
        loading: isLoading,
        position: [
            'panel',
            'preview',
            'relation-modal'
        ],
        /**
     * Disabled when:
     *  - currently if you're cloning a document we don't support publish & clone at the same time.
     *  - the form is submitting
     *  - the active tab is the published tab
     *  - the document is already published & not modified
     *  - the document is being created & not modified
     *  - the user doesn't have the permission to publish
     */ disabled: isCloning || isSubmitting || isLoadingDraftRelations || activeTab === 'published' || !modified && isDocumentPublished || !modified && !document?.documentId || !canPublish,
        label: formatMessage({
            id: 'app.utils.publish',
            defaultMessage: 'Publish'
        }),
        onClick: async ()=>{
            await performPublish();
        },
        dialog: hasDraftRelations ? {
            type: 'dialog',
            variant: 'danger',
            footer: null,
            title: formatMessage({
                id: translations.getTranslation(`popUpwarning.warning.bulk-has-draft-relations.title`),
                defaultMessage: 'Confirmation'
            }),
            content: formatMessage({
                id: translations.getTranslation(`popUpwarning.warning.bulk-has-draft-relations.message`),
                defaultMessage: 'This entry is related to {count, plural, one {# draft entry} other {# draft entries}}. Publishing it could leave broken links in your app.'
            }, {
                count: totalDraftRelations
            }),
            onConfirm: async ()=>{
                await performPublish();
            }
        } : undefined
    };
};
PublishAction.type = 'publish';
PublishAction.position = [
    'panel',
    'preview',
    'relation-modal'
];
const UpdateAction = ({ activeTab, documentId, model, collectionType })=>{
    const dispatchGuidedTour = strapiAdmin.useGuidedTour('UpdateAction', (s)=>s.dispatch);
    const navigate = reactRouterDom.useNavigate();
    const { toggleNotification } = strapiAdmin.useNotification();
    const { _unstableFormatValidationErrors: formatValidationErrors } = strapiAdmin.useAPIErrorHandler();
    const cloneMatch = reactRouterDom.useMatch(router.CLONE_PATH);
    const isCloning = cloneMatch !== null;
    const { formatMessage } = reactIntl.useIntl();
    const { create, update, clone, isLoading } = useDocumentActions.useDocumentActions();
    const { currentDocument: { components } } = useDocumentContext.useDocumentContext('UpdateAction');
    const [{ rawQuery }] = strapiAdmin.useQueryParams();
    const onPreview = Preview.usePreviewContext('UpdateAction', (state)=>state.onPreview, false);
    const { getInitialFormValues } = useDocument.useDoc();
    const isSubmitting = strapiAdmin.useForm('UpdateAction', ({ isSubmitting })=>isSubmitting);
    const modified = strapiAdmin.useForm('UpdateAction', ({ modified })=>modified);
    const setSubmitting = strapiAdmin.useForm('UpdateAction', ({ setSubmitting })=>setSubmitting);
    const initialValues = strapiAdmin.useForm('UpdateAction', ({ initialValues })=>initialValues);
    const document = strapiAdmin.useForm('UpdateAction', ({ values })=>values);
    const validate = strapiAdmin.useForm('UpdateAction', (state)=>state.validate);
    const setErrors = strapiAdmin.useForm('UpdateAction', (state)=>state.setErrors);
    const resetForm = strapiAdmin.useForm('UpdateAction', ({ resetForm })=>resetForm);
    const dispatch = RelationModal.useRelationModal('UpdateAction', (state)=>state.dispatch);
    // need to discriminate if the update is coming from a relation modal or in the edit view
    const relationContext = RelationModal.useRelationModal('UpdateAction', ()=>true, false);
    const relationalModalSchema = RelationModal.useRelationModal('UpdateAction', (state)=>state.currentDocument.schema, false);
    const fieldToConnect = RelationModal.useRelationModal('UpdateAction', (state)=>state.state.fieldToConnect, false);
    const fieldToConnectUID = RelationModal.useRelationModal('PublishAction', (state)=>state.state.fieldToConnectUID, false);
    const documentHistory = RelationModal.useRelationModal('UpdateAction', (state)=>state.state.documentHistory, false);
    const rootDocumentMeta = RelationModal.useRelationModal('UpdateAction', (state)=>state.rootDocumentMeta);
    const fromRelationModal = relationContext != undefined;
    const { currentDocumentMeta } = useDocumentContext.useDocumentContext('UpdateAction');
    const [updateDocumentMutation] = documents.useUpdateDocumentMutation();
    const { _unstableFormatAPIError: formatAPIError } = strapiAdmin.useAPIErrorHandler();
    const parentDocumentMetaToUpdate = documentHistory?.at(-2) ?? rootDocumentMeta;
    const parentDocumentData = useDocument.useDocument({
        documentId: parentDocumentMetaToUpdate?.documentId,
        model: parentDocumentMetaToUpdate?.model,
        collectionType: parentDocumentMetaToUpdate?.collectionType,
        params: parentDocumentMetaToUpdate?.params
    }, {
        skip: !parentDocumentMetaToUpdate
    });
    const { schema } = useDocument.useDoc();
    const handleUpdate = React__namespace.useCallback(async ()=>{
        setSubmitting(true);
        try {
            if (!modified) {
                return;
            }
            const { errors } = await validate(true, {
                status: 'draft'
            });
            if (errors) {
                toggleNotification({
                    type: 'danger',
                    message: formatMessage({
                        id: 'content-manager.validation.error',
                        defaultMessage: 'There are validation errors in your document. Please fix them before saving.'
                    })
                });
                return;
            }
            if (isCloning) {
                const res = await clone({
                    model,
                    documentId: cloneMatch.params.origin,
                    params: currentDocumentMeta.params
                }, transformData(document));
                if ('data' in res) {
                    navigate({
                        pathname: `../${res.data.documentId}`,
                        search: rawQuery
                    }, {
                        relative: 'path'
                    });
                } else if ('error' in res && api.isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                    setErrors(formatValidationErrors(res.error));
                }
            } else if (documentId || collectionType === collections.SINGLE_TYPES) {
                const { data: data$1 } = data.handleInvisibleAttributes(transformData(document), {
                    schema: fromRelationModal ? relationalModalSchema : schema,
                    initialValues,
                    components
                });
                const res = await update({
                    collectionType,
                    model,
                    documentId,
                    params: currentDocumentMeta.params
                }, data$1);
                if ('error' in res && api.isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                    setErrors(formatValidationErrors(res.error));
                } else {
                    resetForm();
                }
            } else {
                const { data: data$1 } = data.handleInvisibleAttributes(transformData(document), {
                    schema: fromRelationModal ? relationalModalSchema : schema,
                    initialValues,
                    components
                });
                const res = await create({
                    model,
                    params: currentDocumentMeta.params
                }, data$1);
                if ('data' in res && collectionType !== collections.SINGLE_TYPES) {
                    if (fromRelationModal) {
                        const createdRelation = {
                            documentId: res.data.documentId,
                            collectionType,
                            model,
                            params: currentDocumentMeta.params
                        };
                        /*
             * Update, if needed, the parent relation with the newly published document.
             * Check if in history we have the parent relation otherwise use the
             * rootDocument
             */ if (fieldToConnect && documentHistory && (parentDocumentMetaToUpdate.documentId || parentDocumentMetaToUpdate.collectionType === collections.SINGLE_TYPES)) {
                            const parentDataToUpdate = parentDocumentMetaToUpdate.collectionType === collections.SINGLE_TYPES ? getInitialFormValues() : parentDocumentData.getInitialFormValues();
                            const dataToUpdate = connectRelationToParent(parentDataToUpdate, fieldToConnect, res.data, fieldToConnectUID);
                            try {
                                const updateRes = await updateDocumentMutation({
                                    collectionType: parentDocumentMetaToUpdate.collectionType,
                                    model: parentDocumentMetaToUpdate.model,
                                    documentId: parentDocumentMetaToUpdate.collectionType !== collections.SINGLE_TYPES ? parentDocumentMetaToUpdate.documentId : undefined,
                                    params: parentDocumentMetaToUpdate.params,
                                    data: {
                                        ...dataToUpdate
                                    }
                                });
                                if ('error' in updateRes) {
                                    toggleNotification({
                                        type: 'danger',
                                        message: formatAPIError(updateRes.error)
                                    });
                                    return;
                                }
                            } catch (err) {
                                toggleNotification({
                                    type: 'danger',
                                    message: formatMessage({
                                        id: 'notification.error',
                                        defaultMessage: 'An error occurred'
                                    })
                                });
                                throw err;
                            }
                        }
                        dispatch({
                            type: 'GO_TO_CREATED_RELATION',
                            payload: {
                                document: createdRelation,
                                shouldBypassConfirmation: true
                            }
                        });
                    } else {
                        navigate({
                            pathname: `../${res.data.documentId}`,
                            search: rawQuery
                        }, {
                            replace: true,
                            relative: 'path'
                        });
                    }
                } else if ('error' in res && api.isBaseQueryError(res.error) && res.error.name === 'ValidationError') {
                    setErrors(formatValidationErrors(res.error));
                }
            }
        } finally{
            dispatchGuidedTour({
                type: 'set_completed_actions',
                payload: [
                    strapiAdmin.GUIDED_TOUR_REQUIRED_ACTIONS.contentManager.createContent
                ]
            });
            setSubmitting(false);
            if (onPreview) {
                onPreview();
            }
        }
    }, [
        setSubmitting,
        modified,
        validate,
        isCloning,
        documentId,
        collectionType,
        toggleNotification,
        formatMessage,
        clone,
        model,
        cloneMatch?.params.origin,
        currentDocumentMeta.params,
        document,
        navigate,
        rawQuery,
        setErrors,
        formatValidationErrors,
        update,
        resetForm,
        create,
        fromRelationModal,
        fieldToConnect,
        documentHistory,
        parentDocumentMetaToUpdate,
        dispatch,
        getInitialFormValues,
        parentDocumentData,
        fieldToConnectUID,
        updateDocumentMutation,
        formatAPIError,
        onPreview,
        initialValues,
        schema,
        components,
        relationalModalSchema,
        dispatchGuidedTour
    ]);
    // Auto-save on CMD+S or CMD+Enter on macOS, and CTRL+S or CTRL+Enter on Windows/Linux
    React__namespace.useEffect(()=>{
        const handleKeyDown = (e)=>{
            if (e.key === 'Enter' && (e.metaKey || e.ctrlKey)) {
                e.preventDefault();
                handleUpdate();
            }
        };
        window.addEventListener('keydown', handleKeyDown);
        return ()=>{
            window.removeEventListener('keydown', handleKeyDown);
        };
    }, [
        handleUpdate
    ]);
    return {
        loading: isLoading,
        /**
     * Disabled when:
     * - the form is submitting
     * - the document is not modified & we're not cloning (you can save a clone entity straight away)
     * - the active tab is the published tab
     */ disabled: isSubmitting || !modified && !isCloning || activeTab === 'published',
        label: formatMessage({
            id: 'global.save',
            defaultMessage: 'Save'
        }),
        onClick: handleUpdate,
        position: [
            'panel',
            'preview',
            'relation-modal'
        ]
    };
};
UpdateAction.type = 'update';
UpdateAction.position = [
    'panel',
    'preview',
    'relation-modal'
];
const UNPUBLISH_DRAFT_OPTIONS = {
    KEEP: 'keep',
    DISCARD: 'discard'
};
const UnpublishAction = ({ activeTab, documentId, model, collectionType, document })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { schema } = useDocument.useDoc();
    const canPublish = DocumentRBAC.useDocumentRBAC('UnpublishAction', ({ canPublish })=>canPublish);
    const { unpublish } = useDocumentActions.useDocumentActions();
    const [{ query }] = strapiAdmin.useQueryParams();
    const params = React__namespace.useMemo(()=>api.buildValidParams(query), [
        query
    ]);
    const { toggleNotification } = strapiAdmin.useNotification();
    const [shouldKeepDraft, setShouldKeepDraft] = React__namespace.useState(true);
    const isDocumentModified = document?.status === 'modified';
    const handleChange = (value)=>{
        setShouldKeepDraft(value === UNPUBLISH_DRAFT_OPTIONS.KEEP);
    };
    if (!schema?.options?.draftAndPublish) {
        return null;
    }
    return {
        disabled: !canPublish || activeTab === 'published' || document?.status !== 'published' && document?.status !== 'modified',
        label: formatMessage({
            id: 'app.utils.unpublish',
            defaultMessage: 'Unpublish'
        }),
        icon: /*#__PURE__*/ jsxRuntime.jsx(Icons.Cross, {}),
        onClick: async ()=>{
            /**
       * return if there's no id & we're in a collection type, or the status modified
       * for either collection type because we use a dialog to handle the process in
       * the latter case.
       */ if (!documentId && collectionType !== collections.SINGLE_TYPES || isDocumentModified) {
                if (!documentId) {
                    // This should never, ever, happen.
                    console.error("You're trying to unpublish a document without an id, this is likely a bug with Strapi. Please open an issue.");
                    toggleNotification({
                        message: formatMessage({
                            id: 'content-manager.actions.unpublish.error',
                            defaultMessage: 'An error occurred while trying to unpublish the document.'
                        }),
                        type: 'danger'
                    });
                }
                return;
            }
            await unpublish({
                collectionType,
                model,
                documentId,
                params
            });
        },
        dialog: isDocumentModified ? {
            type: 'dialog',
            title: formatMessage({
                id: 'app.components.ConfirmDialog.title',
                defaultMessage: 'Confirmation'
            }),
            content: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                alignItems: "flex-start",
                direction: "column",
                gap: 6,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                        width: "100%",
                        direction: "column",
                        gap: 2,
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(Icons.WarningCircle, {
                                width: "24px",
                                height: "24px",
                                fill: "danger600"
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                tag: "p",
                                variant: "omega",
                                textAlign: "center",
                                children: formatMessage({
                                    id: 'content-manager.actions.unpublish.dialog.body',
                                    defaultMessage: 'Are you sure?'
                                })
                            })
                        ]
                    }),
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Radio.Group, {
                        defaultValue: UNPUBLISH_DRAFT_OPTIONS.KEEP,
                        name: "discard-options",
                        "aria-label": formatMessage({
                            id: 'content-manager.actions.unpublish.dialog.radio-label',
                            defaultMessage: 'Choose an option to unpublish the document.'
                        }),
                        onValueChange: handleChange,
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Radio.Item, {
                                checked: shouldKeepDraft,
                                value: UNPUBLISH_DRAFT_OPTIONS.KEEP,
                                children: formatMessage({
                                    id: 'content-manager.actions.unpublish.dialog.option.keep-draft',
                                    defaultMessage: 'Keep draft'
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Radio.Item, {
                                checked: !shouldKeepDraft,
                                value: UNPUBLISH_DRAFT_OPTIONS.DISCARD,
                                children: formatMessage({
                                    id: 'content-manager.actions.unpublish.dialog.option.replace-draft',
                                    defaultMessage: 'Replace draft'
                                })
                            })
                        ]
                    })
                ]
            }),
            onConfirm: async ()=>{
                if (!documentId && collectionType !== collections.SINGLE_TYPES) {
                    // This should never, ever, happen.
                    console.error("You're trying to unpublish a document without an id, this is likely a bug with Strapi. Please open an issue.");
                    toggleNotification({
                        message: formatMessage({
                            id: 'content-manager.actions.unpublish.error',
                            defaultMessage: 'An error occurred while trying to unpublish the document.'
                        }),
                        type: 'danger'
                    });
                }
                await unpublish({
                    collectionType,
                    model,
                    documentId,
                    params
                }, !shouldKeepDraft);
            }
        } : undefined,
        variant: 'danger',
        position: [
            'panel',
            'table-row'
        ]
    };
};
UnpublishAction.type = 'unpublish';
UnpublishAction.position = 'panel';
const DiscardAction = ({ activeTab, documentId, model, collectionType, document })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { schema } = useDocument.useDoc();
    const canUpdate = DocumentRBAC.useDocumentRBAC('DiscardAction', ({ canUpdate })=>canUpdate);
    const { discard, isLoading } = useDocumentActions.useDocumentActions();
    const [{ query }] = strapiAdmin.useQueryParams();
    const params = React__namespace.useMemo(()=>api.buildValidParams(query), [
        query
    ]);
    if (!schema?.options?.draftAndPublish) {
        return null;
    }
    return {
        disabled: !canUpdate || activeTab === 'published' || document?.status !== 'modified',
        label: formatMessage({
            id: 'content-manager.actions.discard.label',
            defaultMessage: 'Discard changes'
        }),
        icon: /*#__PURE__*/ jsxRuntime.jsx(Icons.Cross, {}),
        position: [
            'panel',
            'table-row'
        ],
        variant: 'danger',
        dialog: {
            type: 'dialog',
            title: formatMessage({
                id: 'app.components.ConfirmDialog.title',
                defaultMessage: 'Confirmation'
            }),
            content: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                direction: "column",
                gap: 2,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(Icons.WarningCircle, {
                        width: "24px",
                        height: "24px",
                        fill: "danger600"
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                        tag: "p",
                        variant: "omega",
                        textAlign: "center",
                        children: formatMessage({
                            id: 'content-manager.actions.discard.dialog.body',
                            defaultMessage: 'Are you sure?'
                        })
                    })
                ]
            }),
            loading: isLoading,
            onConfirm: async ()=>{
                await discard({
                    collectionType,
                    model,
                    documentId,
                    params
                });
            }
        }
    };
};
DiscardAction.type = 'discard';
DiscardAction.position = 'panel';
const DEFAULT_ACTIONS = [
    PublishAction,
    UpdateAction,
    UnpublishAction,
    DiscardAction
];

exports.DEFAULT_ACTIONS = DEFAULT_ACTIONS;
exports.DocumentActionButton = DocumentActionButton;
exports.DocumentActions = DocumentActions;
exports.DocumentActionsMenu = DocumentActionsMenu;
