// @strapi/admin/dist/admin/ee/admin/src/services/ai.js
'use strict';

var api = require('../../../../admin/src/services/api.js');

const aiService = api.adminApi.injectEndpoints({
    endpoints: (builder)=>({
            getAIUsage: builder.query({
                query: ()=>({
                        method: 'GET',
                        url: `/admin/ai-usage`
                    }),
                providesTags: [
                    'AIUsage'
                ]
            }),
            getAiToken: builder.query({
                query: ()=>({
                        method: 'GET',
                        url: '/admin/ai-token'
                    }),
                transformResponse (res) {
                    return res.data;
                }
            })
        }),
    overrideExisting: true
});
const { useGetAIUsageQuery, useGetAiTokenQuery, useLazyGetAiTokenQuery } = aiService;

exports.useGetAIUsageQuery = useGetAIUsageQuery;
exports.useGetAiTokenQuery = useGetAiTokenQuery;
exports.useLazyGetAiTokenQuery = useLazyGetAiTokenQuery;
