// @abgov/adsp-service-sdk/src/access/createRealmStrategy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRealmStrategy = void 0;
const tslib_1 = require("tslib");
const jwks_rsa_1 = require("jwks-rsa");
const jwt_decode_1 = require("jwt-decode");
const passport_jwt_1 = require("passport-jwt");
const utils_1 = require("../utils");
const resolveRoles_1 = require("./resolveRoles");
const createRealmStrategy = (_a) => tslib_1.__awaiter(void 0, [_a], void 0, function* ({ realm, serviceId, tenantService, accessServiceUrl, ignoreServiceAud, additionalExtractors, }) {
    (0, utils_1.assertAdspId)(serviceId, null, 'service');
    const serviceAud = serviceId.toString();
    const realmIssUrl = new URL(`/auth/realms/${realm}`, accessServiceUrl);
    const realmIss = realmIssUrl.href;
    const realmJwksUrl = new URL(`/auth/realms/${realm}/protocol/openid-connect/certs`, accessServiceUrl);
    const realmJwks = realmJwksUrl.href;
    const tenant = realm !== 'core' ? yield tenantService.getTenantByRealm(realm) : null;
    const verifyCallback = (req, payload, done) => {
        var _a;
        const user = {
            id: payload.sub,
            name: payload.name || payload.preferred_username,
            email: payload.email,
            roles: (0, resolveRoles_1.resolveRoles)(serviceAud, payload),
            tenantId: tenant === null || tenant === void 0 ? void 0 : tenant.id,
            isCore: realm === 'core',
            token: Object.assign(Object.assign({}, payload), { bearer: (_a = req.headers.authorization) === null || _a === void 0 ? void 0 : _a.substring(7) }),
        };
        done(null, user);
    };
    const jwksHandler = (0, jwks_rsa_1.passportJwtSecret)({
        jwksUri: realmJwks,
        cache: true,
    });
    const strategy = new passport_jwt_1.Strategy({
        jwtFromRequest: passport_jwt_1.ExtractJwt.fromExtractors([
            passport_jwt_1.ExtractJwt.fromAuthHeaderAsBearerToken(),
            ...(additionalExtractors || []),
        ]),
        secretOrKeyProvider: (req, token, done) => {
            try {
                // Decode the header and check against expected issuer.
                // This is necessary because jwks-rsa does not cache the no key result and will make a remote request each time.
                const { iss } = (0, jwt_decode_1.default)(token);
                if (iss !== realmIss) {
                    done(null, null);
                }
                else {
                    jwksHandler(req, token, done);
                }
            }
            catch (err) {
                done(err);
            }
        },
        passReqToCallback: true,
        audience: !ignoreServiceAud ? serviceAud : null,
        issuer: realmIss,
    }, verifyCallback);
    return strategy;
});
exports.createRealmStrategy = createRealmStrategy;
