// @abgov/adsp-service-sdk/src/access/createTenantStrategy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTenantStrategy = void 0;
const tslib_1 = require("tslib");
const passport_jwt_1 = require("passport-jwt");
const utils_1 = require("../utils");
const issuerCache_1 = require("./issuerCache");
const keyProvider_1 = require("./keyProvider");
const resolveRoles_1 = require("./resolveRoles");
const createTenantStrategy = ({ serviceId, tenantService, accessServiceUrl, ignoreServiceAud, accessTokenInQuery, additionalExtractors, logger, }) => {
    (0, utils_1.assertAdspId)(serviceId, null, 'service');
    const serviceAud = serviceId.toString();
    const issuerCache = new issuerCache_1.IssuerCache(logger, accessServiceUrl, tenantService);
    const keyProvider = new keyProvider_1.TenantKeyProvider(logger, accessServiceUrl, issuerCache);
    const verifyCallback = (req, payload, done) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        var _a;
        const tenant = yield issuerCache.getTenantByIssuer(payload.iss);
        const user = {
            id: payload.sub,
            name: payload.name || payload.preferred_username,
            email: payload.email,
            roles: (0, resolveRoles_1.resolveRoles)(serviceAud, payload),
            tenantId: tenant === null || tenant === void 0 ? void 0 : tenant.id,
            isCore: false,
            token: Object.assign(Object.assign({}, payload), { bearer: (_a = req.headers.authorization) === null || _a === void 0 ? void 0 : _a.substring(7) }),
        };
        done(null, user, null);
    });
    const extractors = [passport_jwt_1.ExtractJwt.fromAuthHeaderAsBearerToken(), ...(additionalExtractors || [])];
    if (accessTokenInQuery) {
        extractors.push(passport_jwt_1.ExtractJwt.fromUrlQueryParameter('token'));
    }
    const strategy = new passport_jwt_1.Strategy({
        jwtFromRequest: passport_jwt_1.ExtractJwt.fromExtractors(extractors),
        secretOrKeyProvider: keyProvider.keyRequestHandler,
        passReqToCallback: true,
        audience: !ignoreServiceAud ? serviceAud : null,
    }, verifyCallback);
    return strategy;
};
exports.createTenantStrategy = createTenantStrategy;
