// @abgov/adsp-service-sdk/src/access/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTokenProvider = exports.createTenantStrategy = exports.createRealmStrategy = exports.hasRequiredRole = exports.UnauthorizedUserError = exports.AssertCoreRole = exports.AssertRole = exports.isAllowedUser = void 0;
const tokenProvider_1 = require("./tokenProvider");
var assert_1 = require("./assert");
Object.defineProperty(exports, "isAllowedUser", { enumerable: true, get: function () { return assert_1.isAllowedUser; } });
Object.defineProperty(exports, "AssertRole", { enumerable: true, get: function () { return assert_1.AssertRole; } });
Object.defineProperty(exports, "AssertCoreRole", { enumerable: true, get: function () { return assert_1.AssertCoreRole; } });
Object.defineProperty(exports, "UnauthorizedUserError", { enumerable: true, get: function () { return assert_1.UnauthorizedUserError; } });
Object.defineProperty(exports, "hasRequiredRole", { enumerable: true, get: function () { return assert_1.hasRequiredRole; } });
var createRealmStrategy_1 = require("./createRealmStrategy");
Object.defineProperty(exports, "createRealmStrategy", { enumerable: true, get: function () { return createRealmStrategy_1.createRealmStrategy; } });
var createTenantStrategy_1 = require("./createTenantStrategy");
Object.defineProperty(exports, "createTenantStrategy", { enumerable: true, get: function () { return createTenantStrategy_1.createTenantStrategy; } });
const createTokenProvider = ({ realm, serviceId, clientSecret, accessServiceUrl, logger, }) => {
    return new tokenProvider_1.TokenProviderImpl(logger, serviceId, clientSecret, accessServiceUrl, realm);
};
exports.createTokenProvider = createTokenProvider;
