// @abgov/adsp-service-sdk/src/access/issuerCache.js
"use strict";
var _IssuerCache_issuers, _IssuerCache_getIssuer, _IssuerCache_updateIssuers;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssuerCache = void 0;
const tslib_1 = require("tslib");
const NodeCache = require("node-cache");
class IssuerCache {
    constructor(logger, accessServiceUrl, service) {
        this.logger = logger;
        this.accessServiceUrl = accessServiceUrl;
        this.service = service;
        this.LOG_CONTEXT = { context: 'IssuerCache' };
        _IssuerCache_issuers.set(this, new NodeCache({
            stdTTL: 3600,
            useClones: false,
        }));
        _IssuerCache_getIssuer.set(this, (tenant) => {
            if (!(tenant === null || tenant === void 0 ? void 0 : tenant.realm)) {
                return null;
            }
            const tenantIssuerUrl = new URL(`/auth/realms/${tenant.realm}`, this.accessServiceUrl);
            return tenantIssuerUrl.href;
        });
        _IssuerCache_updateIssuers.set(this, () => tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Updating issuers...'`, this.LOG_CONTEXT);
            const tenants = yield this.service.getTenants();
            if (tenants) {
                tenants.forEach((t) => {
                    const iss = tslib_1.__classPrivateFieldGet(this, _IssuerCache_getIssuer, "f").call(this, t);
                    if (iss) {
                        tslib_1.__classPrivateFieldGet(this, _IssuerCache_issuers, "f").set(iss, t);
                        this.logger.debug(`Caching issuer '${iss}' for tenant '${t.id}'`, this.LOG_CONTEXT);
                    }
                });
                this.logger.info(`Updated issuers.`, this.LOG_CONTEXT);
            }
            else {
                this.logger.warn(`No tenants found during update of issuers.`, this.LOG_CONTEXT);
            }
        }));
        this.getTenantByIssuer = (issuer) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            let tenant = tslib_1.__classPrivateFieldGet(this, _IssuerCache_issuers, "f").get(issuer);
            if (!tenant) {
                try {
                    yield tslib_1.__classPrivateFieldGet(this, _IssuerCache_updateIssuers, "f").call(this);
                    tenant = tslib_1.__classPrivateFieldGet(this, _IssuerCache_issuers, "f").get(issuer);
                }
                catch (err) {
                    this.logger.error(`Encountered error on updating issuers. ${err}`);
                }
            }
            return tenant;
        });
    }
}
exports.IssuerCache = IssuerCache;
_IssuerCache_issuers = new WeakMap(), _IssuerCache_getIssuer = new WeakMap(), _IssuerCache_updateIssuers = new WeakMap();
