// @abgov/adsp-service-sdk/src/access/tokenProvider.js
"use strict";
var _TokenProviderImpl_token, _TokenProviderImpl_expiry;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenProviderImpl = void 0;
const tslib_1 = require("tslib");
const axios_1 = require("axios");
const utils_1 = require("../utils");
class TokenProviderImpl {
    constructor(logger, serviceId, clientSecret, accessServiceUrl, realm = 'core') {
        this.logger = logger;
        this.serviceId = serviceId;
        this.clientSecret = clientSecret;
        this.accessServiceUrl = accessServiceUrl;
        this.realm = realm;
        this.LOG_CONTEXT = { context: 'TokenProvider' };
        _TokenProviderImpl_token.set(this, void 0);
        _TokenProviderImpl_expiry.set(this, void 0);
        this.getAccessToken = () => tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (tslib_1.__classPrivateFieldGet(this, _TokenProviderImpl_token, "f") && Date.now() < tslib_1.__classPrivateFieldGet(this, _TokenProviderImpl_expiry, "f")) {
                this.logger.debug(`Using existing access token...'`, this.LOG_CONTEXT);
                return tslib_1.__classPrivateFieldGet(this, _TokenProviderImpl_token, "f");
            }
            else {
                return yield this.retrieveAccessToken();
            }
        });
        (0, utils_1.assertAdspId)(serviceId, null, 'service');
    }
    retrieveAccessToken() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const authUrl = new URL(`/auth/realms/${this.realm}/protocol/openid-connect/token`, this.accessServiceUrl);
            this.logger.debug(`Requesting access token from ${authUrl}...'`, this.LOG_CONTEXT);
            try {
                const { data } = yield axios_1.default.post(authUrl.href, new URLSearchParams({
                    grant_type: 'client_credentials',
                    client_id: this.serviceId.toString(),
                    client_secret: this.clientSecret,
                }));
                // Store token and expiry (with 60 sec dead-band)
                tslib_1.__classPrivateFieldSet(this, _TokenProviderImpl_token, data.access_token, "f");
                tslib_1.__classPrivateFieldSet(this, _TokenProviderImpl_expiry, Date.now() + (data.expires_in - 60) * 1000, "f");
                this.logger.debug(`Retrieved and cached access token.`, this.LOG_CONTEXT);
                return data.access_token;
            }
            catch (err) {
                this.logger.error(`Error encountered retrieving access token. ${err}`, this.LOG_CONTEXT);
                throw err;
            }
        });
    }
}
exports.TokenProviderImpl = TokenProviderImpl;
_TokenProviderImpl_token = new WeakMap(), _TokenProviderImpl_expiry = new WeakMap();
tslib_1.__decorate([
    (0, utils_1.LimitToOne)(),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Promise)
], TokenProviderImpl.prototype, "retrieveAccessToken", null);
