// @strapi/admin/dist/admin/admin/src/pages/Home/HomePage.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var Overview = require('../../components/GuidedTour/Overview.js');
var Layout = require('../../components/Layouts/Layout.js');
var PageHelpers = require('../../components/PageHelpers.js');
var WidgetHelpers = require('../../components/WidgetHelpers.js');
require('../../services/admin.js');
var useEnterprise = require('../../hooks/useEnterprise.js');
require('../../../../ee/admin/src/services/ai.js');
var Auth = require('../../features/Auth.js');
var StrapiApp = require('../../features/StrapiApp.js');
var Tracking = require('../../features/Tracking.js');
var FreeTrialEndedModal = require('./components/FreeTrialEndedModal.js');
var FreeTrialWelcomeModal = require('./components/FreeTrialWelcomeModal.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const WidgetRoot = ({ title, icon = icons.PuzzlePiece, children, link, uid })=>{
    const { trackUsage } = Tracking.useTracking();
    const { formatMessage } = reactIntl.useIntl();
    const id = React__namespace.useId();
    const Icon = icon;
    const handleClickOnLink = ()=>{
        trackUsage('didOpenHomeWidgetLink', {
            widgetUID: uid
        });
    };
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
        width: "100%",
        hasRadius: true,
        direction: "column",
        alignItems: "flex-start",
        background: "neutral0",
        borderColor: "neutral150",
        shadow: "tableShadow",
        tag: "section",
        gap: 4,
        padding: 6,
        "aria-labelledby": id,
        children: [
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                direction: "row",
                gap: 2,
                justifyContent: "space-between",
                width: "100%",
                tag: "header",
                children: [
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                        gap: 2,
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(Icon, {
                                fill: "neutral500",
                                "aria-hidden": true
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                textColor: "neutral500",
                                variant: "sigma",
                                tag: "h2",
                                id: id,
                                children: formatMessage(title)
                            })
                        ]
                    }),
                    link && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                        tag: reactRouterDom.Link,
                        variant: "omega",
                        textColor: "primary600",
                        style: {
                            textDecoration: 'none'
                        },
                        textAlign: "right",
                        to: link.href,
                        onClick: handleClickOnLink,
                        children: formatMessage(link.label)
                    })
                ]
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                width: "100%",
                height: "261px",
                overflow: "auto",
                tag: "main",
                children: children
            })
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * UnstableHomePageCe
 * -----------------------------------------------------------------------------------------------*/ const WidgetComponent = ({ component })=>{
    const [loadedComponent, setLoadedComponent] = React__namespace.useState(null);
    React__namespace.useEffect(()=>{
        const loadComponent = async ()=>{
            const resolvedComponent = await component();
            setLoadedComponent(()=>resolvedComponent);
        };
        loadComponent();
    }, [
        component
    ]);
    const Component = loadedComponent;
    if (!Component) {
        return /*#__PURE__*/ jsxRuntime.jsx(WidgetHelpers.Widget.Loading, {});
    }
    return /*#__PURE__*/ jsxRuntime.jsx(Component, {});
};
/* -------------------------------------------------------------------------------------------------
 * HomePageCE
 * -----------------------------------------------------------------------------------------------*/ const HomePageCE = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const user = Auth.useAuth('HomePageCE', (state)=>state.user);
    const displayName = user?.firstname ?? user?.username ?? user?.email;
    const getAllWidgets = StrapiApp.useStrapiApp('UnstableHomepageCe', (state)=>state.widgets.getAll);
    const checkUserHasPermissions = Auth.useAuth('WidgetRoot', (state)=>state.checkUserHasPermissions);
    const [filteredWidgets, setFilteredWidgets] = React__namespace.useState([]);
    const [loading, setLoading] = React__namespace.useState(true);
    React__namespace.useEffect(()=>{
        const checkWidgetsPermissions = async ()=>{
            const allWidgets = getAllWidgets();
            const authorizedWidgets = await Promise.all(allWidgets.map(async (widget)=>{
                if (!widget.permissions || widget.permissions.length === 0) return true;
                const matchingPermissions = await checkUserHasPermissions(widget.permissions);
                return matchingPermissions.length >= widget.permissions.length;
            }));
            setFilteredWidgets(allWidgets.filter((_, i)=>authorizedWidgets[i]));
            setLoading(false);
        };
        checkWidgetsPermissions();
    }, [
        checkUserHasPermissions,
        getAllWidgets
    ]);
    return /*#__PURE__*/ jsxRuntime.jsx(Layout.Layouts.Root, {
        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Main, {
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Title, {
                    children: formatMessage({
                        id: 'HomePage.head.title',
                        defaultMessage: 'Homepage'
                    })
                }),
                /*#__PURE__*/ jsxRuntime.jsx(Layout.Layouts.Header, {
                    title: formatMessage({
                        id: 'HomePage.header.title',
                        defaultMessage: 'Hello {name}'
                    }, {
                        name: displayName
                    }),
                    subtitle: formatMessage({
                        id: 'HomePage.header.subtitle',
                        defaultMessage: 'Welcome to your administration panel'
                    })
                }),
                /*#__PURE__*/ jsxRuntime.jsx(FreeTrialWelcomeModal.FreeTrialWelcomeModal, {}),
                /*#__PURE__*/ jsxRuntime.jsx(FreeTrialEndedModal.FreeTrialEndedModal, {}),
                /*#__PURE__*/ jsxRuntime.jsx(Layout.Layouts.Content, {
                    children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                        direction: "column",
                        alignItems: "stretch",
                        gap: 8,
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(Overview.GuidedTourHomepageOverview, {}),
                            loading ? /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                                position: "absolute",
                                top: 0,
                                left: 0,
                                right: 0,
                                bottom: 0,
                                children: /*#__PURE__*/ jsxRuntime.jsx(PageHelpers.Page.Loading, {})
                            }) : /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Root, {
                                gap: 5,
                                children: filteredWidgets.map((widget)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                                        col: 6,
                                        s: 12,
                                        children: /*#__PURE__*/ jsxRuntime.jsx(WidgetRoot, {
                                            title: widget.title,
                                            icon: widget.icon,
                                            link: widget.link,
                                            uid: widget.uid,
                                            children: /*#__PURE__*/ jsxRuntime.jsx(WidgetComponent, {
                                                component: widget.component
                                            })
                                        })
                                    }, widget.uid))
                            })
                        ]
                    })
                })
            ]
        })
    });
};
/* -------------------------------------------------------------------------------------------------
 * HomePage
 * -----------------------------------------------------------------------------------------------*/ const HomePage = ()=>{
    const Page = useEnterprise.useEnterprise(HomePageCE, // eslint-disable-next-line import/no-cycle
    async ()=>(await Promise.resolve().then(function () { return require('../../../../ee/admin/src/pages/HomePage.js'); })).HomePageEE);
    // block rendering until the EE component is fully loaded
    if (!Page) {
        return null;
    }
    return /*#__PURE__*/ jsxRuntime.jsx(Page, {});
};

exports.HomePage = HomePage;
exports.HomePageCE = HomePageCE;
exports.WidgetRoot = WidgetRoot;
