// @strapi/admin/dist/server/server/src/controllers/admin.js
'use strict';

var path = require('path');
var fp = require('lodash/fp');
var _ = require('lodash');
var fse = require('fs-extra');
var utils = require('@strapi/utils');
var tsUtils = require('@strapi/typescript-utils');
var projectSettings = require('../validation/project-settings.js');
var index = require('../utils/index.js');

const { isUsingTypeScript } = tsUtils;
/**
 * A set of functions called "actions" for `Admin`
 */ var admin = {
    // TODO very temporary to check the switch ee/ce
    // When removing this we need to update the /admin/src/index.js file
    // whe,re we set the strapi.window.isEE value
    // NOTE: admin/ee/server overrides this controller, and adds the EE features
    // This returns an empty feature list for CE
    async getProjectType () {
        const flags = strapi.config.get('admin.flags', {});
        const ai = strapi.config.get('admin.ai', {});
        return {
            data: {
                isEE: false,
                features: [],
                flags,
                ai
            }
        };
    },
    async init () {
        let uuid = strapi.config.get('uuid', false);
        const hasAdmin = await index.getService('user').exists();
        const { menuLogo, authLogo } = await index.getService('project-settings').getProjectSettings();
        // set to null if telemetryDisabled flag not avaialble in package.json
        const telemetryDisabled = strapi.config.get('packageJsonStrapi.telemetryDisabled', null);
        if (telemetryDisabled !== null && telemetryDisabled === true) {
            uuid = false;
        }
        return {
            data: {
                uuid,
                hasAdmin,
                menuLogo: menuLogo ? menuLogo.url : null,
                authLogo: authLogo ? authLogo.url : null
            }
        };
    },
    async getProjectSettings () {
        return index.getService('project-settings').getProjectSettings();
    },
    async updateProjectSettings (ctx) {
        const { request: { files, body } } = ctx;
        const projectSettingsService = index.getService('project-settings');
        await projectSettings.validateUpdateProjectSettings(body);
        await projectSettings.validateUpdateProjectSettingsFiles(files);
        const formatedFiles = await projectSettingsService.parseFilesData(files);
        await projectSettings.validateUpdateProjectSettingsImagesDimensions(formatedFiles);
        return projectSettingsService.updateProjectSettings({
            ...body,
            ...formatedFiles
        });
    },
    async telemetryProperties (ctx) {
        // If the telemetry is disabled, ignore the request and return early
        if (strapi.telemetry.isDisabled) {
            ctx.status = 204;
            return;
        }
        const useTypescriptOnServer = await isUsingTypeScript(strapi.dirs.app.root);
        const useTypescriptOnAdmin = await isUsingTypeScript(path.join(strapi.dirs.app.root, 'src', 'admin'));
        const isHostedOnStrapiCloud = utils.env('STRAPI_HOSTING', null) === 'strapi.cloud';
        const aiLicenseKey = utils.env('STRAPI_ADMIN_AI_LICENSE', undefined);
        const numberOfAllContentTypes = _.size(strapi.contentTypes);
        const numberOfComponents = _.size(strapi.components);
        const getNumberOfDynamicZones = ()=>{
            return fp.pipe(fp.map('attributes'), fp.flatMap(fp.values), // @ts-expect-error lodash types
            fp.sumBy(fp.propEq('type', 'dynamiczone')))(strapi.contentTypes);
        };
        return {
            data: {
                useTypescriptOnServer,
                useTypescriptOnAdmin,
                isHostedOnStrapiCloud,
                aiLicenseKey,
                numberOfAllContentTypes,
                numberOfComponents,
                numberOfDynamicZones: getNumberOfDynamicZones()
            }
        };
    },
    async information () {
        const currentEnvironment = strapi.config.get('environment');
        const autoReload = strapi.config.get('autoReload', false);
        const strapiVersion = strapi.config.get('info.strapi', null);
        const dependencies = strapi.config.get('info.dependencies', {});
        const projectId = strapi.config.get('uuid', null);
        const nodeVersion = process.version;
        const communityEdition = !strapi.EE;
        const useYarn = await fse.exists(path.join(process.cwd(), 'yarn.lock'));
        return {
            data: {
                currentEnvironment,
                autoReload,
                strapiVersion,
                dependencies,
                projectId,
                nodeVersion,
                communityEdition,
                useYarn
            }
        };
    },
    async plugins (ctx) {
        const enabledPlugins = strapi.config.get('enabledPlugins');
        // List of core plugins that are always enabled,
        // and so it's not necessary to display them in the plugins list
        const CORE_PLUGINS = [
            'content-manager',
            'content-type-builder',
            'email',
            'upload',
            'i18n',
            'content-releases',
            'review-workflows'
        ];
        const plugins = Object.entries(enabledPlugins).filter(([key])=>!CORE_PLUGINS.includes(key)).map(([key, plugin])=>({
                name: plugin.info.name || key,
                displayName: plugin.info.displayName || plugin.info.name || key,
                description: plugin.info.description || '',
                packageName: plugin.info.packageName
            }));
        ctx.send({
            plugins
        });
    },
    async licenseTrialTimeLeft () {
        const data = await strapi.ee.getTrialEndDate({
            strapi
        });
        return data;
    },
    async getGuidedTourMeta (ctx) {
        const isFirstSuperAdminUser = await index.getService('user').isFirstSuperAdminUser(ctx.state.user.id);
        return {
            data: {
                isFirstSuperAdminUser,
                schemas: strapi.contentTypes
            }
        };
    }
};

module.exports = admin;
