// @strapi/admin/dist/server/ee/server/src/controllers/admin.js
'use strict';

var fp = require('lodash/fp');
var utils = require('@strapi/utils');
var index = require('../utils/index.js');

var admin = {
    // NOTE: Overrides CE admin controller
    async getProjectType () {
        const flags = strapi.config.get('admin.flags', {});
        const ai = strapi.config.get('admin.ai', {});
        try {
            return {
                data: {
                    isEE: strapi.EE,
                    isTrial: strapi.ee.isTrial,
                    features: strapi.ee.features.list(),
                    flags,
                    type: strapi.ee.type,
                    ai
                }
            };
        } catch (err) {
            return {
                data: {
                    isEE: false,
                    features: [],
                    flags
                }
            };
        }
    },
    async licenseLimitInformation () {
        const permittedSeats = strapi.ee.seats;
        let shouldNotify = false;
        let licenseLimitStatus = null;
        let enforcementUserCount;
        const currentActiveUserCount = await index.getService('user').getCurrentActiveUserCount();
        const eeDisabledUsers = await index.getService('seat-enforcement').getDisabledUserList();
        if (eeDisabledUsers) {
            enforcementUserCount = currentActiveUserCount + eeDisabledUsers.length;
        } else {
            enforcementUserCount = currentActiveUserCount;
        }
        if (!fp.isNil(permittedSeats) && enforcementUserCount > permittedSeats) {
            shouldNotify = true;
            licenseLimitStatus = 'OVER_LIMIT';
        }
        if (!fp.isNil(permittedSeats) && enforcementUserCount === permittedSeats) {
            shouldNotify = true;
            licenseLimitStatus = 'AT_LIMIT';
        }
        const data = {
            enforcementUserCount,
            currentActiveUserCount,
            permittedSeats,
            shouldNotify,
            shouldStopCreate: fp.isNil(permittedSeats) ? false : currentActiveUserCount >= permittedSeats,
            licenseLimitStatus,
            isHostedOnStrapiCloud: utils.env('STRAPI_HOSTING', null) === 'strapi.cloud',
            aiLicenseKey: utils.env('STRAPI_ADMIN_AI_LICENSE'),
            type: strapi.ee.type,
            isTrial: strapi.ee.isTrial,
            features: strapi.ee.features.list() ?? []
        };
        return {
            data
        };
    }
};

module.exports = admin;
