// @abgov/adsp-service-sdk/src/healthCheck.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHealthCheck = exports.checkServiceHealth = void 0;
const tslib_1 = require("tslib");
const axios_1 = require("axios");
const NodeCache = require("node-cache");
const utils_1 = require("./utils");
const checkServiceHealth = (logger, healthUrl) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    try {
        const { status } = yield axios_1.default.get(healthUrl.href, { timeout: 1000 });
        return status >= 200 && status < 300;
    }
    catch (err) {
        logger.debug(`Encountered error in health check. ${err}`, { context: 'checkHealth' });
        return false;
    }
});
exports.checkServiceHealth = checkServiceHealth;
const createHealthCheck = (logger, accessServiceUrl, directoryUrl, directory, exclude) => {
    // This cache is a dead-band on health check requests.
    const resultCache = new NodeCache({ stdTTL: 120 });
    return () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        let results = resultCache.get('health');
        if (!results) {
            const accessHealthUrl = new URL('/auth/realms/core/.well-known/openid-configuration', accessServiceUrl);
            const accessHealth = exclude.access || (yield (0, exports.checkServiceHealth)(logger, accessHealthUrl));
            const directoryHealthUrl = new URL('/', directoryUrl);
            const directoryHealth = exclude.directory || (yield (0, exports.checkServiceHealth)(logger, directoryHealthUrl));
            const tenantUrl = yield directory.getServiceUrl((0, utils_1.adspId) `urn:ads:platform:tenant-service`);
            const tenantHealthUrl = new URL('/', tenantUrl);
            const tenantHealth = exclude.tenant || (yield (0, exports.checkServiceHealth)(logger, tenantHealthUrl));
            const configurationUrl = yield directory.getServiceUrl((0, utils_1.adspId) `urn:ads:platform:configuration-service`);
            const configurationHealthUrl = new URL('/', configurationUrl);
            const configurationHealth = exclude.configuration || (yield (0, exports.checkServiceHealth)(logger, configurationHealthUrl));
            const eventUrl = yield directory.getServiceUrl((0, utils_1.adspId) `urn:ads:platform:event-service`);
            const eventHealthUrl = new URL('/', eventUrl);
            const eventHealth = exclude.event || (yield (0, exports.checkServiceHealth)(logger, eventHealthUrl));
            results = {
                access: accessHealth,
                directory: directoryHealth,
                tenant: tenantHealth,
                configuration: configurationHealth,
                event: eventHealth,
            };
            Object.getOwnPropertyNames(exclude).forEach((excludeKey) => {
                if (exclude[excludeKey]) {
                    delete results[excludeKey];
                }
            });
            resultCache.set('health', results);
        }
        return results;
    });
};
exports.createHealthCheck = createHealthCheck;
